### ISSUE TITLE
Investigate support for transient NameID

### ISSUE THREAD

**Issue by NicolasLiampotis** [eosc-kc/keycloak#39]
See also https://issues.redhat.com/browse/KEYCLOAK-5657

**Comment by cgeorgilakis** [comment]
In current Keycloak ( version 12.0.2) admin can select in Identity Provider configuration to identify and track external SAML Identity provider (Principal Type field)  by this values : 

- Subject NameID ( default)
- Attribute [Name]
- Attribute [Friedly Name]

In the attribute selection, admin must provide Principal Attribute. 

Transient selection can be added to ui. Moreover, we need to not allow admin to select transient together with Subject NameID in Principal Type. Is this enough for transient NameID IdPs? In this jira issue Keycloak team propose this as solution together with controlling AllowCreate in AuthnRequest. I do not know what this mean. Now AllowCreate  is always set to true.

Related to xml parsing, I set Principal Type field to default value (Subject NameID) only in Identity Provider creation. In Identity Provider update from xml, Principal Type does not change. I believe that we also need here a change. From SAML metadata, how I can get Attribute Name (or Attribute Friedly Name) and set appropriate values? Otherwise we will have a transient IdP with Subject NameID and many different users will be created in Keycloak for the same SAML Identity Provider user.

**Comment by NicolasLiampotis** [comment]
> In current Keycloak ( version 12.0.2) admin can select in Identity Provider configuration to identify and track external SAML Identity provider (Principal Type field) by this values :
> 
> * Subject NameID ( default)
> * Attribute [Name]
> * Attribute [Friedly Name]
> 
> In the attribute selection, admin must provide Principal Attribute.
> 
> Transient selection can be added to ui. Moreover, we need to not allow admin to select transient together with Subject NameID in Principal Type. Is this enough for transient NameID IdPs? In this jira issue Keycloak team propose this as solution together with controlling AllowCreate in AuthnRequest. I do not know what this mean. Now AllowCreate is always set to true.

I agree that Subject NameID should not be allowed as Principal Type when the supported NameID is transient.

The `AllowCreate` (see Section 3.4.1.1 in https://docs.oasis-open.org/security/saml/v2.0/saml-core-2.0-os.pdf) is not related to transient NameID. Still, it would be useful to be able to override the current behaviour which is not consistent with the SAML core spec where `AllowCreate` defaults to `false`. I've created issue #43 for tracking this.

> Related to xml parsing, I set Principal Type field to default value (Subject NameID) only in Identity Provider creation. In Identity Provider update from xml, Principal Type does not change. I believe that we also need here a change. From SAML metadata, how I can get Attribute Name (or Attribute Friedly Name) and set appropriate values? Otherwise we will have a transient IdP with Subject NameID and many different users will be created in Keycloak for the same SAML Identity Provider user.
This is tricky because if an IdP only supports transient NameIDs it is not straightforward how to determine which attribute to use as the Principal Type. This is one of the reasons we need to support multiple candidate attributes as user identifiers - see #40.

> 
> Finally, for SAMl federation in common exported SAML SP Metadata this value is taken from SAML Federation configuration.
I'm confused. Isn't the NameID extracted from each IdP's metadata? How does the SAML federation configuration affect the NameID policy of each individual IdP if the IdP already advertises one or more supported NameID formats?

**Comment by NicolasLiampotis** [comment]
Upstream PR: https://github.com/keycloak/keycloak/pull/7806
